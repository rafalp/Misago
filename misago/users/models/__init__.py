from .rank import Rank
from .online import Online
from .user import AnonymousUser, User, UsernameChange
from .activityranking import ActivityRanking
from .avatar import Avatar
from .audittrail import AuditTrail
from .avatargallery import AvatarGallery
from .ban import Ban, BanCache
from .datadownload import DataDownload
from .deleteduser import DeletedUser
