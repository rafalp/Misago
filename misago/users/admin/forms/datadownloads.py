from django import forms
from django.contrib.auth import get_user_model
from django.db.models import Q
from django.utils.translation import pgettext, pgettext_lazy

from ...models import DataDownload
from ...utils import hash_email

User = get_user_model()


class RequestDataDownloadsForm(forms.Form):
    user_identifiers = forms.CharField(
        label=pgettext_lazy("admin data download request form", "Usernames or emails"),
        help_text=pgettext_lazy(
            "admin data download request form",
            "Enter every item in new line. Duplicates will be ignored. This field is case insensitive. Depending on site configuration and amount of data to archive it may take up to few days for requests to complete. E-mail will notification will be sent to every user once their download is ready.",
        ),
        widget=forms.Textarea,
    )

    def clean_user_identifiers(self):
        user_identifiers = self.cleaned_data["user_identifiers"].lower().splitlines()
        user_identifiers = list(filter(bool, user_identifiers))
        user_identifiers = list(set(user_identifiers))

        if len(user_identifiers) > 20:
            raise forms.ValidationError(
                pgettext(
                    "admin data download request form",
                    "You may not enter more than 20 items at a single time (You have entered %(show_value)s).",
                )
                % {"show_value": len(user_identifiers)}
            )

        return user_identifiers

    def clean(self):
        data = super().clean()

        if data.get("user_identifiers"):
            username_match = Q(slug__in=data["user_identifiers"])
            email_match = Q(email_hash__in=map(hash_email, data["user_identifiers"]))

            data["users"] = list(User.objects.filter(username_match | email_match))

            if len(data["users"]) != len(data["user_identifiers"]):
                raise forms.ValidationError(
                    pgettext(
                        "admin data download request form",
                        "One or more specified users could not be found.",
                    )
                )

        return data


class FilterDataDownloadsForm(forms.Form):
    status = forms.ChoiceField(
        label=pgettext_lazy("admin data download requests filter form", "Status"),
        required=False,
        choices=DataDownload.STATUS_CHOICES,
    )
    user = forms.CharField(
        label=pgettext_lazy("admin data download requests filter form", "User"),
        required=False,
    )
    requested_by = forms.CharField(
        label=pgettext_lazy("admin data download requests filter form", "Requested by"),
        required=False,
    )

    def filter_queryset(self, criteria, queryset):
        if criteria.get("status") is not None:
            queryset = queryset.filter(status=criteria["status"])

        if criteria.get("user"):
            queryset = queryset.filter(user__slug__istartswith=criteria["user"])

        if criteria.get("requested_by"):
            queryset = queryset.filter(
                requester__slug__istartswith=criteria["requested_by"]
            )

        return queryset
