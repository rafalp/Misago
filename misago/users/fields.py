from typing import TYPE_CHECKING

from django import forms
from django.contrib.auth import get_user_model
from django.db.models import QuerySet
from django.utils.translation import npgettext_lazy, pgettext_lazy

from ..forms.widgets import ListInput
from ..core.utils import slugify

if TYPE_CHECKING:
    from ..users.models import User


class UserMultipleChoiceWidget(forms.MultiWidget):
    template_name = "misago/widgets/user_multiple_choice.html"
    subfields = ("default", "noscript")

    def __init__(self, attrs: dict | None = None):
        super().__init__(
            {
                "": ListInput,
                "noscript": forms.TextInput,
            },
            attrs=attrs,
        )

    def get_context(self, name, value, attrs):
        max_choices = attrs.pop("max_choices", None)
        choices = attrs.pop("choices", None)

        # Decompress value manually because MultiWidget never calls decompress for lists
        value = self.decompress_unknown_value(value)

        context = super().get_context(name, value, attrs)
        for i, subwidget in enumerate(context["widget"]["subwidgets"]):
            subwidget_name = self.subfields[i]
            context["widget"][subwidget_name] = subwidget

        del context["widget"]["subwidgets"]

        context["widget"]["max_choices"] = max_choices
        context["widget"]["choices"] = choices

        return context

    def decompress_unknown_value(self, value) -> list:
        if not value or not isinstance(value, (list, tuple)):
            # unknown value, discard
            return [None, None]

        if isinstance(value[0], get_user_model()):
            # Iterable[User], initial value
            return [value, " ".join(user.username for user in value)]

        return value


class UserMultipleChoiceField(forms.MultiValueField):
    widget = UserMultipleChoiceWidget

    def __init__(
        self, *, queryset: QuerySet | None = None, max_choices: int = 5, **kwargs
    ):
        super().__init__(
            fields=(
                UserMultipleChoiceJavaScriptSubField(required=False),
                UserMultipleChoiceNoScriptSubField(required=False),
            ),
            require_all_fields=False,
            **kwargs,
        )

        self.max_choices = max_choices
        self.queryset = queryset or get_user_model().objects

    def _get_max_choices(self) -> int:
        return self._max_choices

    def _set_max_choices(self, max_choices: int):
        self._max_choices = max_choices
        for field in self.fields:
            field.max_choices = max_choices

    max_choices: int = property(_get_max_choices, _set_max_choices)

    def _get_queryset(self) -> QuerySet:
        return self._queryset

    def _set_queryset(self, queryset: QuerySet):
        self._queryset = queryset
        for field in self.fields:
            field.queryset = queryset

    queryset: QuerySet = property(_get_queryset, _set_queryset)

    def get_users_cache(self) -> list["User"] | None:
        for field in self.fields:
            if users_cache := getattr(field, "users_cache", None):
                return users_cache
        return None

    def compress(self, data_list: tuple[list[str], str]) -> list["User"]:
        if data_list:
            value, value_noscript = data_list
            # Prioritize the noscript value because its only present with the JS disabled
            return value_noscript or value
        return None

    def get_bound_field(self, form, field_name):
        return UserMultipleChoiceBoundField(form, self, field_name)


class UserMultipleChoiceSubField(forms.Field):
    default_error_messages = {
        "invalid_choice": pgettext_lazy(
            "user multiple choice field error", "One or more users not found: %(value)s"
        ),
        "max_choices": npgettext_lazy(
            "user multiple choice field error",
            "Select no more than %(max)d user.",
            "Select no more than %(max)d users.",
            "max",
        ),
    }

    def __init__(
        self, queryset: QuerySet | None = None, max_choices: int = 5, **kwargs
    ):
        self.max_choices = max_choices
        self.queryset = queryset

        super().__init__(**kwargs)

    def get_users(self, queryset: QuerySet, usernames: list[str]) -> list["User"]:
        raise NotImplementedError()

    def validate_max_choices(self, value_length: int):
        if value_length > self.max_choices:
            raise forms.ValidationError(
                self.error_messages["max_choices"],
                code="max_choices",
                params={"max": self.max_choices},
            )


class UserNotFound:
    id = None
    username: str

    __slots__ = ("username",)

    def __init__(self, username: str):
        self.username = username


class UserMultipleChoiceJavaScriptSubField(UserMultipleChoiceSubField):
    widget = ListInput

    def to_python(self, value) -> list[str]:
        return value or []

    def clean(self, usernames: list[str]) -> list["User"]:
        usernames = self.to_python(usernames)

        if usernames:
            value = self.get_users(self.queryset, usernames)
            self.users_cache = value
            self.validate_max_choices(len(value))
        else:
            value = []

        self.validate(value)
        self.run_validators(value)
        return value

    def get_users(self, queryset: QuerySet, usernames: list[str]) -> list["User"]:
        slugs: dict[str, str] = {}
        for username in usernames:
            slug = slugify(username)
            slugs[slug] = username

        users = list(queryset.filter(slug__in=slugs)[: self.max_choices + 16])
        users_dict = {user.slug: user for user in users}

        ordered_users = []
        for slug, username in slugs.items():
            if slug in users_dict:
                ordered_users.append(users_dict[slug])
            else:
                # Return "blank" user item to appear in the UI
                ordered_users.append(UserNotFound(username))

        return ordered_users

    def validate(self, value: list["User"]):
        invalid_choices: list[str] = [user.username for user in value if not user.id]
        if invalid_choices:
            raise forms.ValidationError(
                self.error_messages["invalid_choice"],
                code="invalid_choice",
                params={"value": ", ".join(invalid_choices)},
            )


class UserMultipleChoiceNoScriptSubField(UserMultipleChoiceSubField):
    def to_python(self, value) -> list[str]:
        if not value:
            return []

        final_value = []
        for item in value.split():
            if "," in item:
                final_value += [i.strip() for i in item.split(",")]
            else:
                final_value.append(item)
        return [item for item in final_value if item]

    def clean(self, usernames: list[str]) -> list["User"]:
        usernames = self.to_python(usernames)

        if usernames:
            self.validate_max_choices(len(usernames))
            value = self.get_users(self.queryset, usernames)
        else:
            value = []

        self.validate(value, usernames)
        self.run_validators(value)
        return value

    def get_users(self, queryset: QuerySet, usernames: list[str]) -> list["User"]:
        slugs: set[str] = set(slugify(username) for username in usernames)
        return list(queryset.filter(slug__in=slugs)[: self.max_choices + 1])

    def validate(self, value: list["User"], usernames: list[str]):
        slugs: set[str] = set(user.slug for user in value)
        invalid_choices: list[str] = []
        for username in usernames:
            username_slug = slugify(username)
            if username_slug not in slugs:
                invalid_choices.append(username)

        if invalid_choices:
            raise forms.ValidationError(
                self.error_messages["invalid_choice"],
                code="invalid_choice",
                params={"value": ", ".join(invalid_choices)},
            )


class UserMultipleChoiceBoundField(forms.BoundField):
    @property
    def max_choices(self) -> int:
        return self.field.max_choices

    def build_widget_attrs(self, attrs: dict, widget: forms.Widget | None = None):
        attrs = super().build_widget_attrs(attrs, widget)
        attrs["max_choices"] = self.max_choices
        attrs["choices"] = self.field.get_users_cache() or self.initial or []
        return attrs
