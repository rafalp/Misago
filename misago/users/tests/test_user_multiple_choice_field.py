import pytest
from django.forms import Form
from django.http import QueryDict

from ..fields import UserMultipleChoiceField
from ..models import User


@pytest.fixture
def users(admin, moderator, user, other_user):
    return (admin, moderator, user, other_user)


def fuzz_html_user_ids(html_str: str, users: list[User]) -> str:
    for new_id, user in enumerate(users, 1):
        html_str = html_str.replace(f'="{user.id}"', f'="{new_id}"')
    return html_str


def test_user_multiple_choice_field_renders_without_data(snapshot):
    class TestForm(Form):
        users = UserMultipleChoiceField()

    form = TestForm()

    field_html = str(form["users"])
    assert "users" in field_html
    assert field_html == snapshot


def test_user_multiple_choice_field_renders_with_form_initial_data(users, snapshot):
    class TestForm(Form):
        users = UserMultipleChoiceField()

    form = TestForm(initial={"users": users})

    field_html = str(form["users"])
    assert "users" in field_html

    for user in users:
        assert user.username in field_html
        assert f'value="{user.username}"' in field_html

    usernames = " ".join(user.username for user in users)
    assert f'value="{usernames}"' in field_html

    assert fuzz_html_user_ids(field_html, users) == snapshot


def test_user_multiple_choice_field_renders_with_field_initial_data(users, snapshot):
    field_initial = users

    class TestForm(Form):
        users = UserMultipleChoiceField(initial=field_initial)

    form = TestForm()

    field_html = str(form["users"])
    assert "users" in field_html

    for user in users:
        assert user.username in field_html
        assert f'value="{user.username}"' in field_html

    usernames = " ".join(user.username for user in users)
    assert f'value="{usernames}"' in field_html

    assert fuzz_html_user_ids(field_html, users) == snapshot


def test_user_multiple_choice_field_renders_without_query_data_for_field(
    users, snapshot
):
    class TestForm(Form):
        users = UserMultipleChoiceField()

    form = TestForm(QueryDict("&".join(f"users={user.username}" for user in users)))

    field_html = str(form["users"])
    assert "users" in field_html

    # Bound but not yet cleaned data is useless for rendering purposes
    for user in users:
        assert user.username not in field_html
        assert f'value="{user.username}"' not in field_html

    usernames = " ".join(user.username for user in users)
    assert f'value="{usernames}"' not in field_html
    assert 'value=""' in field_html

    assert fuzz_html_user_ids(field_html, users) == snapshot


def test_user_multiple_choice_field_renders_with_query_data_for_noscript_field(
    users, snapshot
):
    class TestForm(Form):
        users = UserMultipleChoiceField()

    usernames = " ".join(user.username for user in users)

    form = TestForm(QueryDict(f"users_noscript={usernames}"))

    field_html = str(form["users"])
    assert "users" in field_html

    for user in users:
        assert user.username in field_html
        assert f'value="{user.username}"' not in field_html

    assert f'value="{usernames}"' in field_html
    assert fuzz_html_user_ids(field_html, users) == snapshot


def test_user_multiple_choice_field_renders_with_cleaned_query_data_for_field(
    users, snapshot
):
    class TestForm(Form):
        users = UserMultipleChoiceField()

    form = TestForm(QueryDict("&".join(f"users={user.username}" for user in users)))
    assert form.is_valid()

    field_html = str(form["users"])
    assert "users" in field_html

    for user in users:
        assert user.username in field_html
        assert f'value="{user.username}"' in field_html

    usernames = " ".join(user.username for user in users)
    assert f'value="{usernames}"' not in field_html
    assert 'value=""' in field_html

    assert fuzz_html_user_ids(field_html, users) == snapshot


def test_user_multiple_choice_field_renders_with_cleaned_query_data_for_noscript_field(
    users, snapshot
):
    class TestForm(Form):
        users = UserMultipleChoiceField()

    usernames = " ".join(user.username for user in users)

    form = TestForm(QueryDict(f"users_noscript={usernames}"))
    assert form.is_valid()

    field_html = str(form["users"])
    assert "users" in field_html

    for user in users:
        assert user.username in field_html
        assert f'value="{user.username}"' not in field_html

    assert f'value="{usernames}"' in field_html

    assert fuzz_html_user_ids(field_html, users) == snapshot


def test_user_multiple_choice_field_renders_with_cleaned_query_data_for_field(
    users, snapshot
):
    class TestForm(Form):
        users = UserMultipleChoiceField()

    form = TestForm(QueryDict("&".join(f"users={user.username}" for user in users)))
    assert form.is_valid()

    field_html = str(form["users"])
    assert "users" in field_html

    for user in users:
        assert user.username in field_html
        assert f'value="{user.username}"' in field_html

    usernames = " ".join(user.username for user in users)
    assert f'value="{usernames}"' not in field_html
    assert 'value=""' in field_html

    assert fuzz_html_user_ids(field_html, users) == snapshot


def test_user_multiple_choice_field_prioritizes_noscript_field(admin, other_user):
    class TestForm(Form):
        users = UserMultipleChoiceField()

    form = TestForm(
        QueryDict(f"users={other_user.username}&users_noscript={admin.username}")
    )
    assert form.is_valid()
    assert form.cleaned_data["users"] == [admin]


def test_user_multiple_choice_field_prioritizes_noscript_field(admin, other_user):
    class TestForm(Form):
        users = UserMultipleChoiceField()

    form = TestForm(
        QueryDict(f"users={other_user.username}&users_noscript={admin.username}")
    )
    assert form.is_valid()
    assert form.cleaned_data["users"] == [admin]


def test_user_multiple_choice_field_validates_value_max_choices(
    admin, moderator, other_user
):
    class TestForm(Form):
        users = UserMultipleChoiceField(max_choices=2)

    form = TestForm(
        QueryDict(
            f"users={admin.username}&users={moderator.username}&users={other_user.username}"
        )
    )
    assert not form.is_valid()
    assert form.errors == {"users": ["Select no more than 2 users."]}


def test_user_multiple_choice_field_validates_noscript_value_max_choices(
    admin, moderator, other_user
):
    class TestForm(Form):
        users = UserMultipleChoiceField(max_choices=2)

    form = TestForm(
        QueryDict(
            f"users_noscript={admin.username} {moderator.username} {other_user.username}"
        )
    )
    assert not form.is_valid()
    assert form.errors == {"users": ["Select no more than 2 users."]}


def test_user_multiple_choice_field_fails_to_validate_nonexisting_users(
    admin, snapshot
):
    class TestForm(Form):
        users = UserMultipleChoiceField(max_choices=10)

    form = TestForm(QueryDict(f"users=John&users={admin.username}&users=Doe"))
    assert not form.is_valid()
    assert form.errors == {"users": ["One or more users not found: John, Doe"]}

    field_html = str(form["users"])
    assert "users" in field_html
    assert fuzz_html_user_ids(field_html, [admin]) == snapshot


def test_user_multiple_choice_field_fails_to_validate_noscript_nonexisting_users(
    admin, snapshot
):
    class TestForm(Form):
        users = UserMultipleChoiceField(max_choices=10)

    form = TestForm(QueryDict(f"users_noscript=John {admin.username} Doe"))
    assert not form.is_valid()
    assert form.errors == {"users": ["One or more users not found: John, Doe"]}

    field_html = str(form["users"])
    assert "users" in field_html
    assert fuzz_html_user_ids(field_html, [admin]) == snapshot
