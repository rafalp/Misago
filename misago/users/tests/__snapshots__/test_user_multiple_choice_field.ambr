# serializer version: 1
# name: test_user_multiple_choice_field_fails_to_validate_nonexisting_users
  '''
  <ul class="form-chips-control d-js-flex" maxchoices="10" m-user-multiple-choice>
      
        
          <li m-user-multiple-choice-item>
            <input type="hidden" name="users" value="John">
            <button class="form-chips-control-item-not-found" type="button" aria-label="Remove this choice">
              <span class="form-chips-control-item-not-found-icon">
                <span class="material-icon">person_outline</span>
              </span>
              <span class="form-chips-control-item-name">
                John
              </span>
              <span class="form-chips-control-item-close-icon">
                <span class="material-icon">clear</span>
              </span>
            </button>
          </li>
        
      
        
          <li m-user-multiple-choice-item m-user-id="1">
            <input type="hidden" name="users" value="Admin_User">
            <button class="form-chips-control-item" type="button" aria-label="Remove this choice">
              <img class="form-chips-control-item-image" src="http://dummyimage.com/100/100" alt="">
              <span class="form-chips-control-item-name">
                Admin_User
              </span>
              <span class="form-chips-control-item-close-icon">
                <span class="material-icon">clear</span>
              </span>
            </button>
          </li>
        
      
        
          <li m-user-multiple-choice-item>
            <input type="hidden" name="users" value="Doe">
            <button class="form-chips-control-item-not-found" type="button" aria-label="Remove this choice">
              <span class="form-chips-control-item-not-found-icon">
                <span class="material-icon">person_outline</span>
              </span>
              <span class="form-chips-control-item-name">
                Doe
              </span>
              <span class="form-chips-control-item-close-icon">
                <span class="material-icon">clear</span>
              </span>
            </button>
          </li>
        
      
      <li>
        <input type="text" class="form-chips-control-input" id="id_users_0" m-user-multiple-choice-input>
      </li>
    </ul>
    <template id="m-user-multiple-choice-template">
      <li m-user-multiple-choice-item>
        <input type="hidden" name="users" value="">
        <button class="form-chips-control-item" type="button" aria-label="Remove this choice">
          <img class="form-chips-control-item-image" src="" alt="">
          <span class="form-chips-control-item-name">
            <slot name="username"></slot>
          </span>
          <span class="form-chips-control-item-close-icon">
            <span class="material-icon">clear</span>
          </span>
        </button>
      </li>
    </template>
  
  
    <noscript>
      <input type="text" class="form-control" id="id_users_1" name="users_noscript" value="" >
    </noscript>
  '''
# ---
# name: test_user_multiple_choice_field_fails_to_validate_noscript_nonexisting_users
  '''
  <ul class="form-chips-control d-js-flex" maxchoices="10" m-user-multiple-choice>
      
      <li>
        <input type="text" class="form-chips-control-input" id="id_users_0" m-user-multiple-choice-input>
      </li>
    </ul>
    <template id="m-user-multiple-choice-template">
      <li m-user-multiple-choice-item>
        <input type="hidden" name="users" value="">
        <button class="form-chips-control-item" type="button" aria-label="Remove this choice">
          <img class="form-chips-control-item-image" src="" alt="">
          <span class="form-chips-control-item-name">
            <slot name="username"></slot>
          </span>
          <span class="form-chips-control-item-close-icon">
            <span class="material-icon">clear</span>
          </span>
        </button>
      </li>
    </template>
  
  
    <noscript>
      <input type="text" class="form-control" id="id_users_1" name="users_noscript" value="John Admin_User Doe" >
    </noscript>
  '''
# ---
# name: test_user_multiple_choice_field_renders_with_cleaned_query_data_for_field
  '''
  <ul class="form-chips-control d-js-flex" maxchoices="5" m-user-multiple-choice>
      
        
          <li m-user-multiple-choice-item m-user-id="1">
            <input type="hidden" name="users" value="Admin_User">
            <button class="form-chips-control-item" type="button" aria-label="Remove this choice">
              <img class="form-chips-control-item-image" src="http://dummyimage.com/100/100" alt="">
              <span class="form-chips-control-item-name">
                Admin_User
              </span>
              <span class="form-chips-control-item-close-icon">
                <span class="material-icon">clear</span>
              </span>
            </button>
          </li>
        
      
        
          <li m-user-multiple-choice-item m-user-id="2">
            <input type="hidden" name="users" value="Moderator">
            <button class="form-chips-control-item" type="button" aria-label="Remove this choice">
              <img class="form-chips-control-item-image" src="http://dummyimage.com/100/100" alt="">
              <span class="form-chips-control-item-name">
                Moderator
              </span>
              <span class="form-chips-control-item-close-icon">
                <span class="material-icon">clear</span>
              </span>
            </button>
          </li>
        
      
        
          <li m-user-multiple-choice-item m-user-id="3">
            <input type="hidden" name="users" value="User">
            <button class="form-chips-control-item" type="button" aria-label="Remove this choice">
              <img class="form-chips-control-item-image" src="http://dummyimage.com/100/100" alt="">
              <span class="form-chips-control-item-name">
                User
              </span>
              <span class="form-chips-control-item-close-icon">
                <span class="material-icon">clear</span>
              </span>
            </button>
          </li>
        
      
        
          <li m-user-multiple-choice-item m-user-id="4">
            <input type="hidden" name="users" value="Other_User">
            <button class="form-chips-control-item" type="button" aria-label="Remove this choice">
              <img class="form-chips-control-item-image" src="http://dummyimage.com/100/100" alt="">
              <span class="form-chips-control-item-name">
                Other_User
              </span>
              <span class="form-chips-control-item-close-icon">
                <span class="material-icon">clear</span>
              </span>
            </button>
          </li>
        
      
      <li>
        <input type="text" class="form-chips-control-input" id="id_users_0" m-user-multiple-choice-input>
      </li>
    </ul>
    <template id="m-user-multiple-choice-template">
      <li m-user-multiple-choice-item>
        <input type="hidden" name="users" value="">
        <button class="form-chips-control-item" type="button" aria-label="Remove this choice">
          <img class="form-chips-control-item-image" src="" alt="">
          <span class="form-chips-control-item-name">
            <slot name="username"></slot>
          </span>
          <span class="form-chips-control-item-close-icon">
            <span class="material-icon">clear</span>
          </span>
        </button>
      </li>
    </template>
  
  
    <noscript>
      <input type="text" class="form-control" id="id_users_1" name="users_noscript" value="" >
    </noscript>
  '''
# ---
# name: test_user_multiple_choice_field_renders_with_cleaned_query_data_for_noscript_field
  '''
  <ul class="form-chips-control d-js-flex" maxchoices="5" m-user-multiple-choice>
      
      <li>
        <input type="text" class="form-chips-control-input" id="id_users_0" m-user-multiple-choice-input>
      </li>
    </ul>
    <template id="m-user-multiple-choice-template">
      <li m-user-multiple-choice-item>
        <input type="hidden" name="users" value="">
        <button class="form-chips-control-item" type="button" aria-label="Remove this choice">
          <img class="form-chips-control-item-image" src="" alt="">
          <span class="form-chips-control-item-name">
            <slot name="username"></slot>
          </span>
          <span class="form-chips-control-item-close-icon">
            <span class="material-icon">clear</span>
          </span>
        </button>
      </li>
    </template>
  
  
    <noscript>
      <input type="text" class="form-control" id="id_users_1" name="users_noscript" value="Admin_User Moderator User Other_User" >
    </noscript>
  '''
# ---
# name: test_user_multiple_choice_field_renders_with_field_initial_data
  '''
  <ul class="form-chips-control d-js-flex" maxchoices="5" m-user-multiple-choice>
      
        
          <li m-user-multiple-choice-item m-user-id="1">
            <input type="hidden" name="users" value="Admin_User">
            <button class="form-chips-control-item" type="button" aria-label="Remove this choice">
              <img class="form-chips-control-item-image" src="http://dummyimage.com/100/100" alt="">
              <span class="form-chips-control-item-name">
                Admin_User
              </span>
              <span class="form-chips-control-item-close-icon">
                <span class="material-icon">clear</span>
              </span>
            </button>
          </li>
        
      
        
          <li m-user-multiple-choice-item m-user-id="2">
            <input type="hidden" name="users" value="Moderator">
            <button class="form-chips-control-item" type="button" aria-label="Remove this choice">
              <img class="form-chips-control-item-image" src="http://dummyimage.com/100/100" alt="">
              <span class="form-chips-control-item-name">
                Moderator
              </span>
              <span class="form-chips-control-item-close-icon">
                <span class="material-icon">clear</span>
              </span>
            </button>
          </li>
        
      
        
          <li m-user-multiple-choice-item m-user-id="3">
            <input type="hidden" name="users" value="User">
            <button class="form-chips-control-item" type="button" aria-label="Remove this choice">
              <img class="form-chips-control-item-image" src="http://dummyimage.com/100/100" alt="">
              <span class="form-chips-control-item-name">
                User
              </span>
              <span class="form-chips-control-item-close-icon">
                <span class="material-icon">clear</span>
              </span>
            </button>
          </li>
        
      
        
          <li m-user-multiple-choice-item m-user-id="4">
            <input type="hidden" name="users" value="Other_User">
            <button class="form-chips-control-item" type="button" aria-label="Remove this choice">
              <img class="form-chips-control-item-image" src="http://dummyimage.com/100/100" alt="">
              <span class="form-chips-control-item-name">
                Other_User
              </span>
              <span class="form-chips-control-item-close-icon">
                <span class="material-icon">clear</span>
              </span>
            </button>
          </li>
        
      
      <li>
        <input type="text" class="form-chips-control-input" id="id_users_0" m-user-multiple-choice-input>
      </li>
    </ul>
    <template id="m-user-multiple-choice-template">
      <li m-user-multiple-choice-item>
        <input type="hidden" name="users" value="">
        <button class="form-chips-control-item" type="button" aria-label="Remove this choice">
          <img class="form-chips-control-item-image" src="" alt="">
          <span class="form-chips-control-item-name">
            <slot name="username"></slot>
          </span>
          <span class="form-chips-control-item-close-icon">
            <span class="material-icon">clear</span>
          </span>
        </button>
      </li>
    </template>
  
  
    <noscript>
      <input type="text" class="form-control" id="id_users_1" name="users_noscript" value="Admin_User Moderator User Other_User" >
    </noscript>
  '''
# ---
# name: test_user_multiple_choice_field_renders_with_form_initial_data
  '''
  <ul class="form-chips-control d-js-flex" maxchoices="5" m-user-multiple-choice>
      
        
          <li m-user-multiple-choice-item m-user-id="1">
            <input type="hidden" name="users" value="Admin_User">
            <button class="form-chips-control-item" type="button" aria-label="Remove this choice">
              <img class="form-chips-control-item-image" src="http://dummyimage.com/100/100" alt="">
              <span class="form-chips-control-item-name">
                Admin_User
              </span>
              <span class="form-chips-control-item-close-icon">
                <span class="material-icon">clear</span>
              </span>
            </button>
          </li>
        
      
        
          <li m-user-multiple-choice-item m-user-id="2">
            <input type="hidden" name="users" value="Moderator">
            <button class="form-chips-control-item" type="button" aria-label="Remove this choice">
              <img class="form-chips-control-item-image" src="http://dummyimage.com/100/100" alt="">
              <span class="form-chips-control-item-name">
                Moderator
              </span>
              <span class="form-chips-control-item-close-icon">
                <span class="material-icon">clear</span>
              </span>
            </button>
          </li>
        
      
        
          <li m-user-multiple-choice-item m-user-id="3">
            <input type="hidden" name="users" value="User">
            <button class="form-chips-control-item" type="button" aria-label="Remove this choice">
              <img class="form-chips-control-item-image" src="http://dummyimage.com/100/100" alt="">
              <span class="form-chips-control-item-name">
                User
              </span>
              <span class="form-chips-control-item-close-icon">
                <span class="material-icon">clear</span>
              </span>
            </button>
          </li>
        
      
        
          <li m-user-multiple-choice-item m-user-id="4">
            <input type="hidden" name="users" value="Other_User">
            <button class="form-chips-control-item" type="button" aria-label="Remove this choice">
              <img class="form-chips-control-item-image" src="http://dummyimage.com/100/100" alt="">
              <span class="form-chips-control-item-name">
                Other_User
              </span>
              <span class="form-chips-control-item-close-icon">
                <span class="material-icon">clear</span>
              </span>
            </button>
          </li>
        
      
      <li>
        <input type="text" class="form-chips-control-input" id="id_users_0" m-user-multiple-choice-input>
      </li>
    </ul>
    <template id="m-user-multiple-choice-template">
      <li m-user-multiple-choice-item>
        <input type="hidden" name="users" value="">
        <button class="form-chips-control-item" type="button" aria-label="Remove this choice">
          <img class="form-chips-control-item-image" src="" alt="">
          <span class="form-chips-control-item-name">
            <slot name="username"></slot>
          </span>
          <span class="form-chips-control-item-close-icon">
            <span class="material-icon">clear</span>
          </span>
        </button>
      </li>
    </template>
  
  
    <noscript>
      <input type="text" class="form-control" id="id_users_1" name="users_noscript" value="Admin_User Moderator User Other_User" >
    </noscript>
  '''
# ---
# name: test_user_multiple_choice_field_renders_with_query_data_for_noscript_field
  '''
  <ul class="form-chips-control d-js-flex" maxchoices="5" m-user-multiple-choice>
      
      <li>
        <input type="text" class="form-chips-control-input" id="id_users_0" m-user-multiple-choice-input>
      </li>
    </ul>
    <template id="m-user-multiple-choice-template">
      <li m-user-multiple-choice-item>
        <input type="hidden" name="users" value="">
        <button class="form-chips-control-item" type="button" aria-label="Remove this choice">
          <img class="form-chips-control-item-image" src="" alt="">
          <span class="form-chips-control-item-name">
            <slot name="username"></slot>
          </span>
          <span class="form-chips-control-item-close-icon">
            <span class="material-icon">clear</span>
          </span>
        </button>
      </li>
    </template>
  
  
    <noscript>
      <input type="text" class="form-control" id="id_users_1" name="users_noscript" value="Admin_User Moderator User Other_User" >
    </noscript>
  '''
# ---
# name: test_user_multiple_choice_field_renders_without_data
  '''
  <ul class="form-chips-control d-js-flex" maxchoices="5" m-user-multiple-choice>
      
      <li>
        <input type="text" class="form-chips-control-input" id="id_users_0" m-user-multiple-choice-input>
      </li>
    </ul>
    <template id="m-user-multiple-choice-template">
      <li m-user-multiple-choice-item>
        <input type="hidden" name="users" value="">
        <button class="form-chips-control-item" type="button" aria-label="Remove this choice">
          <img class="form-chips-control-item-image" src="" alt="">
          <span class="form-chips-control-item-name">
            <slot name="username"></slot>
          </span>
          <span class="form-chips-control-item-close-icon">
            <span class="material-icon">clear</span>
          </span>
        </button>
      </li>
    </template>
  
  
    <noscript>
      <input type="text" class="form-control" id="id_users_1" name="users_noscript" value="" >
    </noscript>
  '''
# ---
# name: test_user_multiple_choice_field_renders_without_query_data_for_field
  '''
  <ul class="form-chips-control d-js-flex" maxchoices="5" m-user-multiple-choice>
      
      <li>
        <input type="text" class="form-chips-control-input" id="id_users_0" m-user-multiple-choice-input>
      </li>
    </ul>
    <template id="m-user-multiple-choice-template">
      <li m-user-multiple-choice-item>
        <input type="hidden" name="users" value="">
        <button class="form-chips-control-item" type="button" aria-label="Remove this choice">
          <img class="form-chips-control-item-image" src="" alt="">
          <span class="form-chips-control-item-name">
            <slot name="username"></slot>
          </span>
          <span class="form-chips-control-item-close-icon">
            <span class="material-icon">clear</span>
          </span>
        </button>
      </li>
    </template>
  
  
    <noscript>
      <input type="text" class="form-control" id="id_users_1" name="users_noscript" value="" >
    </noscript>
  '''
# ---
